docs: Custom Applications multiple permissions (#2806)
commercetools/merchant-center-application-kit: website/src/content/development/permissions.mdx

title: Permissions
---

> As we learned in [OAuth Scopes and user permissions](/concepts/oauth-scopes-and-user-permissions), each Custom Application has a unique pair of default user permissions at its disposal: "view" and "manage."

Reviewer: I think we can simply mention the additional groups here in the intro.

```suggestion
As we learned in [OAuth Scopes and user permissions](/concepts/oauth-scopes-and-user-permissions), each Custom Application has a unique pair of default user permissions at its disposal: "view" and "manage." Additionally, more granular permission groups can be used to cover more strict business requirements.
```

The values of the user permissions are derived from the application `entryPointUriPath`.

When developing a Custom Application you might want to enforce these user permissions in some parts of your application. For example, performing certain actions like creating, updating, or deleting a resource should only be possible if the user has "manage" permission.

> Furthermore, more granular additional permissions can optionally be used, for example, to allow the team access to only certain parts or functionality of the Custom Application.

Reviewer: Can be removed, see suggestion for the intro section.

```suggestion
```

You can then use the `PERMISSIONS` variable to reference the permission in the application code.

<Info>

This feature is available from version `21.21.0` onwards.

When using additional permission groups, the unique group names should be provided to the `entryPointUriPathToPermissionKeys` function to generate the correct permission keys.

The group names can be exported and referenced also in the Custom Application config file.

```js title="constants.js" highlightLines="5-8, 12"
import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';

export const entryPointUriPath = 'channels';

> export const groupNames = {
>   teamA: 'team-a',
>   teamB: 'team-b',
> };
> 
> export const PERMISSIONS = entryPointUriPathToPermissionKeys(
>   entryPointUriPath,
>   [groupNames.teamA, groupNames.teamB]
> );

Reviewer: See other suggestion about the group names

```suggestion
export const groupNames = {
  delivery: 'delivery',
  promotion: 'promotion',
};

export const PERMISSIONS = entryPointUriPathToPermissionKeys(
  entryPointUriPath,
  ['delivery', 'promotion']
);
```

Also, if we provide a TS example, we need to make sure to use the `as const` for the `groupNames`, so that we can pass the list using variable references

```ts
export const groupNames = {
  delivery: 'delivery',
  promotion: 'promotion',
} as const;

export const PERMISSIONS = entryPointUriPathToPermissionKeys(
  entryPointUriPath,
  [groupNames.delivery, groupNames.promotion]
);
```

Author: Thanks! That's a great catch!
:question: I'm still wondering how to approach the TS snippet. Is your point to integrate `@commercetools-docs/gatsby-theme-code-examples` as part of this PR?

Reviewer: No that should probably be a follow up.

```

> In this scenario the `PERMISSIONS` variable contains a `View`, `Manage`, `ViewTeamA`, `ManageTeamA`, `ViewTeamB` and `ManageTeamB` properties, with the values being the computed values based on the `entryPointUriPath` and the provided permission group names:
> 
> * `PERMISSIONS.View`: maps to `ViewChannels`.
> * `PERMISSIONS.Manage`: maps to `ManageChannels`.
> * `PERMISSIONS.ViewTeamA`: maps to `ViewChannelsTeamA`.
> * `PERMISSIONS.ManageTeamA`: maps to `ManageChannelsTeamA`.
> * `PERMISSIONS.ViewTeamB`: maps to `ViewChannelsTeamB`.
> * `PERMISSIONS.ManageTeamB`: maps to `ManageChannelsTeamB`.

Reviewer: ```suggestion
In this scenario the `PERMISSIONS` variable contains a `View`, `Manage`, `ViewDelivery`, `ManageDelivery`, `ViewPromotion` and `ManagePromotion` properties, with the values being the computed values based on the `entryPointUriPath` and the provided permission group names:

* `PERMISSIONS.View`: maps to `ViewChannels`.
* `PERMISSIONS.Manage`: maps to `ManageChannels`.
* `PERMISSIONS.ViewDelivery`: maps to `ViewChannelsDelivery`.
* `PERMISSIONS.ManageDelivery`: maps to `ManageChannelsDelivery`.
* `PERMISSIONS.ViewPromotion`: maps to `ViewChannelsPromotion`.
* `PERMISSIONS.ManagePromotion`: maps to `ManageChannelsPromotion`.
```

You can then use the `PERMISSIONS` variable to reference the permission in the application code.

> <Info>

Reviewer: ```suggestion
<Info>

This feature is available from version `21.21.0` onwards.
```

> To leverage the optional, more granular permissions, a distinctive name has to be introduced for each of the permission groups.

Reviewer: ```suggestion
When using additional permission groups, the unique group names should be provided to the `entryPointUriPathToPermissionKeys` function to generate the correct permission keys.

The group names can be exported and referenced also int he Custom Application config file.
```

```js title="constants.js" highlightLines="5-7, 11"
import { entryPointUriPathToPermissionKeys } from '@commercetools-frontend/application-shell/ssr';

export const entryPointUriPath = 'channels';

> export const teamA = "team-a";
> 
> export const teamB = "team-b";
> 
> export const PERMISSIONS = entryPointUriPathToPermissionKeys(
>   entryPointUriPath,
>   [teamA, teamB]
> );

Reviewer: Nit: maybe we can define the names as an constants object.

```suggestion
export const groupNames = {
  teamA: 'team-a',
  teamB: 'team-b',
};

export const PERMISSIONS = entryPointUriPathToPermissionKeys(
  entryPointUriPath,
  [groupNames.teamA, groupNames.teamB]
);
```
